algorithmic_efficiency/logging_utils.py

```diff
@@ -0,0 +1,492 @@
+  Instead of hardcoding a list of properties to be extracted, we automatically
+  extract any float, int, str, or bool because each workload is different and
+  may change."""
+  workload_properties = {}
+  skip_list = ['param_shapes']
+  keys = [
+      key for key in dir(workload)
+      if not key.startswith('_') and key not in skip_list
+  ]
+  for key in keys:
+    try:
+      attr = getattr(workload, key)
+    except:
+      logging.warn(
+          f'Unable to record workload.{key} information. Continuing without it.'
+      )
+    if _is_primitive_type(attr):
+      workload_properties[f'workload.{key}'] = attr
+  return workload_properties
+
+
+class NoOpRecorder(object):
+  """ This dummy class returns None for all possible function calls.
+
+  This class makes it easy to turn off all functionality by swapping in this
+  class.
+   """
+
+  def no_op(self, *args, **kw):
+    pass
+
+  def __getattr__(self, _):
+    return self.no_op
+
+
+class Recorder:
+  """Save information about the training progress of a workload to disk.
+
+  This class should be instantiated once per workload. Logging files are written
+  to seperate workload specific folders.
+
+  Three files are written to the given "logging_dir" folder:
+  1. "metadata.json" is created at the start of a workload and it includes the
+     datetime, workload name, and system configuration.
+  2. "measurements.csv" is created for each hyperparameter tuning trial and a
+     row is appended for every model evaluation. The information included is
+     loss, accuracy, training step, time elapsed, hparams, workload properties,
+     and hardware utilization.
+  3. "packages.txt" is created at the start of a workload and it includes a
+     list of the currently installed OS and python packages.
+
+  Joining measurement CSVs across workloads or hyperparameter tuning trials is
+  left to users, although a convienence function called "concatenate_csvs()" is
+  provided. The data format of "measurements.csv" is designed to be safe to
+  arbitrarily join CSVs without attribute name conflicts across both workloads
+  and across hyperparameter tuning trials.
+  """
+
+  def __init__(self,
+               workload: spec.Workload,
+               workload_name: str,
+               logging_dir: str,
+               submission_path: str,
+               tuning_ruleset: str,
+               tuning_search_space_path: Optional[str] = None,
+               num_tuning_trials: Optional[int] = None):
+    self.workload_name = workload_name
+    self.workload = workload
+    self.logging_dir = logging_dir
+    self.submission_path = submission_path
+    self.tuning_ruleset = tuning_ruleset
+    self.tuning_search_space_path = tuning_search_space_path
+    self.num_tuning_trials = num_tuning_trials
+    self.status = 'INCOMPLETE'
+    self.last_epoch_evaluated = None
+    self.workload_log_dir = os.path.join(self.logging_dir, self.workload_name)
```

Comment:
I don't mind keeping them as is, but in general if you don't plan on accessing these properties from outside the class, their names should start with a `_`.

Reply:
Fixed. Thanks!